Fix value == undefined and propertyValue != undefined in cookie-utilities.js

francescorn opened the pull request:
#### f7e6c1a5fe40cf9aa797560702f82307c7816837
<pre>
Fix value == undefined and propertyValue != undefined in cookie-utilities.js
<a href="https://bugs.webkit.org/show_bug.cgi?id=274109">https://bugs.webkit.org/show_bug.cgi?id=274109</a>

Reviewed by Tim Nguyen.

Check if propertyValue is undefined and null and the value is undefined or null.
Replace != and !== with === and !== to be more specific, there is no such thing as almost undefined or almost null.
Properties are either null or not null and defined or undefined.

* LayoutTests/http/tests/cookies/resources/cookie-utilities.js:
(createCookie):

Canonical link: <a href="https://commits.webkit.org/278889@main">https://commits.webkit.org/278889@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ec72d32455faaf5d8763424078706a2087c64729

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51679 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30991 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/4041 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54945 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2371 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37356 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/2055 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/42062 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53778 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28641 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44579 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23190 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25937 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1839 "Passed tests") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47897 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1930 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56537 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26800 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1807 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49464 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/28037 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44649 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/48688 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11329 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28934 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27774 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

JosephPecoraro commented on LayoutTests/http/tests/cookies/resources/cookie-utilities.js:
```diff
@@ -12,7 +12,7 @@ function createCookie(name, value, additionalProperties)
         cookie += `; ${propertyName}`;
         let propertyValue = additionalProperties[propertyName];
-        if (propertyValue != undefined)
+        if (propertyValue !== undefined)
```
Typically a library uses ` == undefined` when it wants to check either `null` or `undefined`.

* `null == undefined` => `true`
* `undefined == undefined` => `true`

I don't know anything about this test code, but I wonder:

* Would it make sense for createCookie to allow a cookie value of `null` here?
* Likewise for usages below, I wonder if the library wants to remain resilient if the JSON serialized values from `getCookies` included a `null` value.

Just food for thought. I don't think anything about the change is wrong, just want to make sure some thought was given to the change in behavior to `null` values because I didn't see that described in the commit message.

francescorn replied:
Thank you for bringing up the point. I think we can write instead to be more specific: if (value === undefined || value === null) and if (propertyValue !== undefined && propertyValue !== null).

francescorn committed:
Fix value == undefined and propertyValue != undefined in cookie-utilities.js
https://bugs.webkit.org/show_bug.cgi?id=274109

Reviewed by Tim Nguyen.

Check if propertyValue is undefined and null and the value is undefined or null.
Replace != and !== with === and !== to be more specific, there is no such thing as almost undefined or almost null.
Properties are either null or not null and defined or undefined.

* LayoutTests/http/tests/cookies/resources/cookie-utilities.js:
(createCookie):

Canonical link: https://commits.webkit.org/278889@main
